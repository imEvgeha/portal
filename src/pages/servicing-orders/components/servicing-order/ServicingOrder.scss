.servicing-order{
  display: flex;
  height: 100%;
  &__left {
    display: flex;
    flex-direction: column;
    min-width: 345px;
    border-right: 1px solid #DCDCDC;
    padding: 24px 0 0;
  }
  &__right {
    padding: 24px 20px 0;
  }
}