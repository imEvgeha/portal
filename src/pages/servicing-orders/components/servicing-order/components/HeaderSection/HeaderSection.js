import React, {useState} from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import ArrowLeftIcon from '@atlaskit/icon/glyph/arrow-left';
import {backArrowColor} from '../../../../../legacy/constants/avails/constants';
import {SERVICING_ORDERS} from '../../../../../../ui/elements/nexus-navigation/constants';
import './HeaderSection.scss';
import FulfillmentOrderPanel from '../FulfillmentOrderPanel/FulfillmentOrderPanel';
import FilterSolidIcon from '../../../../../../assets/filter-solid.svg';
import FilterIcon from '../../../../../../assets/filter.svg';
import ServiceOrderFilter from '../FilterSection/ServiceOrderFilter';

const HeaderSection = ({fulfillmentOrders,  orderDetails}) => {
    const [selectedFulfillmentOrder, setSelectedFulfillmentOrder] = useState('');
    const [showFilter, setShowFilter] = useState(false);
    const setSelectedOrder = (id) => setSelectedFulfillmentOrder(id);
    const toggleFilters = () => setShowFilter(!showFilter);

    return (
        <div className='panel-header'>
            <div className='panel-header__title'>
                <div>
                    <Link to={`/${SERVICING_ORDERS.toLowerCase()}`}>
                        <ArrowLeftIcon size='large' primaryColor={backArrowColor} />
                    </Link>
                    <span className='panel-header__title--text'>Servicing Order</span>
                </div>
                <div className='panel-header__filter'>
                    <div onClick={toggleFilters}>
                        {
                            showFilter ? <FilterSolidIcon /> : <FilterIcon />
                        }
                    </div>
                </div>
            </div>
            {
                showFilter && <ServiceOrderFilter orderDetails={orderDetails} />
            }
            <div
                className='panel-header__list'
            >
                {
                    fulfillmentOrders && fulfillmentOrders.map(({fulfillmentOrderId, status, dueDate},index) => (
                        <FulfillmentOrderPanel
                            key={index}
                            id={fulfillmentOrderId}
                            status={status}
                            dueDate={dueDate}
                            selected={selectedFulfillmentOrder === fulfillmentOrderId}
                            setSelectedFulfillmentOrder={setSelectedOrder}
                        />
                        )
                    )
                }
            </div>
        </div>
    );
};

HeaderSection.propTypes = {
    fulfillmentOrders: PropTypes.array,
    orderDetails: PropTypes.object
};
export default HeaderSection;