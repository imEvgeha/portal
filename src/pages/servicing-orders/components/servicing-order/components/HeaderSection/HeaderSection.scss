.panel-header{
  &__title {
    padding-left: 10px;
    svg {
      cursor: pointer;
    }
    &--text{
      font-size: 29px;
      color: #42526E;
      margin-left: 5px;
    }
  }
  &__list {
    display: flex;
    flex-direction: column;
    overflow-y: auto;
    overflow-x: hidden;
    max-height: 100%;
    & > :first-child {
      border-top: 1px solid #DCDCDC;
    }
  }
}