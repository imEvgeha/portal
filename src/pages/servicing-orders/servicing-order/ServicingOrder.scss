@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/mixins';

.servicing-order {
    display: flex;
    min-height: 100%;

    &__left {
        display: flex;
        flex-direction: column;
        min-width: 390px;
        max-width: 400px;
        border-right: 1px solid map-get($light-neutral-colors, light-neutral-2);
    }

    &__right {
        @include padding(6, 5, 0);
        width: 100%;
        background-color: white;
        overflow: auto;
    }
}
