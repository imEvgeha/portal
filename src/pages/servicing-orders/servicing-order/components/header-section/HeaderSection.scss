@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/mixins.scss';
.panel-header {
    &__title {
        @include margin(6, 0);
        display: flex;
        justify-content: space-around;
        svg {
            cursor: pointer;
        }

        &--text {
            @include heading(h800);
            @include margin(0, 0, 0, 2);
            color: map-get($dark-neutral-colors, dark-neutral-5);
        }

        &--section {
            display: flex;
            align-items: center;
        }

        &--open {
            @include margin(6, 0, 3, 0);
        }
    }

    &__list {
        display: flex;
        flex-direction: column;
        overflow-y: auto;
        overflow-x: hidden;
        height: 60vh;

        & > :first-child {
            border-top: 1px solid map-get($light-neutral-colors, light-neutral-2);
        }
    }

    &__filter {
        display: flex;
        align-items: center;

        svg {
            @include addSpacing(width, 5);
            @include addSpacing(height, 5);
            fill: map-get($dark-neutral-colors, dark-neutral-5);
        }
    }
}
