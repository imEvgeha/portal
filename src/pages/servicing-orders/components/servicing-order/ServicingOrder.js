import React, {useEffect, useState} from 'react';
import HeaderSection from './components/HeaderSection/HeaderSection';
import FulfillmentOrder from './components/FulfillmentOrder/FulfillmentOrder';
import './ServicingOrder.scss';
import {servicingOrdersService} from '../../servicingOrdersService';
import {get} from 'lodash';

const ServicingOrder = ({match}) => {
    const [serviceOrder, setServiceOrder] = useState({});
    const [selectedFulfillmentOrderID, setSelectedFulfillmentOrderID] = useState('');
    const [selectedOrder, setSelectedOrder] = useState({});

    useEffect(() => {
        setSelectedOrder(get(serviceOrder, 'fulfillmentOrders', []).find(s=> s && s.fulfillmentOrderId === selectedFulfillmentOrderID) || {});
    }, [serviceOrder, selectedFulfillmentOrderID]);

    useEffect(() => {
        servicingOrdersService.getServicingOrderById(match.params.id) .then(res => {
            const servicingOrder = res['servicingOrder'];
            setServiceOrder(servicingOrder.data || {});
        });
    }, []);

    return (
        <div className='servicing-order'>
            <div className='servicing-order__left'>
                { 
                    serviceOrder && Array.isArray(serviceOrder.fulfillmentOrders) && (
                    <HeaderSection
                        orderDetails={serviceOrder}
                        setSelectedFulfillmentOrder={setSelectedFulfillmentOrderID}
                        selectedFulfillmentOrder={selectedFulfillmentOrderID}
                    />
                    )
                }
            </div>
            <div className='servicing-order__right'>
                <FulfillmentOrder selectedFulfillmentOrder={selectedOrder} />
            </div>
        </div>
    );
};

ServicingOrder.propTypes = {};

ServicingOrder.defaultProps = {};

export default ServicingOrder;
