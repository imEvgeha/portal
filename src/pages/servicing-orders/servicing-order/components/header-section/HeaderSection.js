import ArrowLeftIcon from '@atlaskit/icon/glyph/arrow-left';
import {get} from 'lodash';
import moment from 'moment';
import PropTypes from 'prop-types';
import React, {useState} from 'react';
import {Link} from 'react-router-dom';
import FilterSolidIcon from '../../../../../assets/filter-solid.svg';
import FilterIcon from '../../../../../assets/filter.svg';
import {SERVICING_ORDERS} from '../../../../../ui/elements/nexus-navigation/constants';
import {getValidDate} from '../../../../../util/utils';
import {backArrowColor} from '../../../../legacy/constants/avails/constants';
import {SORT_DIRECTION} from '../filter-section/constants';
import ServiceOrderFilter from '../filter-section/ServiceOrderFilter';
import Constants from '../fulfillment-order/constants';
import FulfillmentOrderPanel from '../fulfillment-order-panel/FulfillmentOrderPanel';
import './HeaderSection.scss';

const HeaderSection = ({orderDetails, handleFulfillmentOrderChange, selectedFulfillmentOrder}) => {
    const [showFilter, setShowFilter] = useState(true);
    const [filter, setFilter] = useState({ value: 'All', label: 'All' });
    const [dueDateSortDirection, setDueDateSortDirection] = useState(SORT_DIRECTION[0]);

    const toggleFilters = () => setShowFilter(!showFilter);
    const getFilteredList = () => {
        let filteredList = [];
        if (orderDetails.fulfillmentOrders && Array.isArray(orderDetails.fulfillmentOrders)) {
            if (!filter || filter.value === 'All') {
                filteredList = orderDetails.fulfillmentOrders;
            } else {
                filteredList = orderDetails.fulfillmentOrders.filter(
                    item => item.status === filter.value
                );
            }
        }
        return filteredList;
    };

    const sortByDueDate = (prevFulfillmentOrder, currFulfillmentOrder) => {
        const getMomentDueDate = fulfillmentOrder => {
            return moment(get(fulfillmentOrder, 'definition.dueDate'));
        };

        const prevFulfillmentOrderDueDate = getMomentDueDate(prevFulfillmentOrder);
        const currFulfillmentOrderDueDate = getMomentDueDate(currFulfillmentOrder);
        const diff = prevFulfillmentOrderDueDate.diff(currFulfillmentOrderDueDate);

        switch (dueDateSortDirection.value) {
            case 'NONE':
                return 0;

            case 'ASCENDING':
                return diff;

            case 'DESCENDING':
                return -diff;
        }
    };

    return (
        <div className="panel-header">
            <div className="panel-header__title">
                <div className="panel-header__title--section">
                    <Link to={`/${SERVICING_ORDERS.toLowerCase()}`}>
                        <ArrowLeftIcon size="large" primaryColor={backArrowColor} />
                    </Link>
                    <h1>Servicing Order</h1>
                </div>
                <div className="panel-header__filter">
                    <div onClick={toggleFilters}>
                        {showFilter ? <FilterSolidIcon /> : <FilterIcon />}
                    </div>
                </div>
            </div>
            {showFilter && (
                <ServiceOrderFilter
                    orderDetails={orderDetails || {}}
                    filter={filter}
                    setFilter={setFilter}
                    dueDateSortDirection={dueDateSortDirection}
                    setDueDateSortDirection={setDueDateSortDirection}
                />
            )}
            <div className="panel-header__list">
                {getFilteredList()
                    .sort(sortByDueDate)
                    .map(({id, external_id, status, definition: {dueDate} = {}}, index) => (
                        <FulfillmentOrderPanel
                            key={index}
                            id={id}
                            external_id={external_id}
                            status={status}
                            dueDate={getValidDate(dueDate)}
                            selected={selectedFulfillmentOrder === id}
                            handleFulfillmentOrderChange={handleFulfillmentOrderChange}
                        />
                    ))}
            </div>
        </div>
    );
};

HeaderSection.propTypes = {
    orderDetails: PropTypes.object.isRequired,
    handleFulfillmentOrderChange: PropTypes.func,
    selectedFulfillmentOrder: PropTypes.string
};

HeaderSection.defaultProps = {
    handleFulfillmentOrderChange: () => null,
    selectedFulfillmentOrder: ''
};
export default HeaderSection;
