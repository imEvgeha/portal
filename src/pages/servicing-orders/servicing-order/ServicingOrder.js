import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {get} from 'lodash';
import {URL} from '../../../util/Common';
import {sortByDateFn} from '../../../util/date-time/DateTimeUtils';
import {servicingOrdersService} from '../servicingOrdersService';
import FulfillmentOrder from './components/fulfillment-order/FulfillmentOrder';
import HeaderSection from './components/header-section/HeaderSection';
import ServicesTable from './components/services-table/ServicesTable';
import SourcesTable from './components/sources-table/SourcesTable';
import {
    prepareRowData,
    showLoading,
    fetchAssetInfo,
    getBarCodes,
    populateAssetInfo,
    removeNulls,
} from './components/sources-table/util';
import './ServicingOrder.scss';

const ServicingOrder = ({match}) => {
    const [serviceOrder, setServiceOrder] = useState({});
    const [selectedFulfillmentOrderID, setSelectedFulfillmentOrderID] = useState('');
    const [selectedOrder, setSelectedOrder] = useState({});
    const [selectedSource, setSelectedSource] = useState();
    const [lastOrder, setLastOrder] = useState({});

    // this piece of state is used for when a service is updated in the services table
    const [updatedServices, setUpdatedServices] = useState({});

    useEffect(() => {
        const order =
            get(serviceOrder, 'fulfillmentOrders', []).find(s => s && s.id === selectedFulfillmentOrderID) || {};
        setSelectedOrder(order);
        setLastOrder(order);
    }, [serviceOrder, selectedFulfillmentOrderID]);

    const fetchFulfillmentOrders = async servicingOrder => {
        if (servicingOrder.so_number) {
            try {
                if (URL.isLocalOrDevOrQA()) {
                    let {
                        fulfillmentOrders,
                        servicingOrderItems,
                    } = await servicingOrdersService.getFulfilmentOrdersForServiceOrder(servicingOrder.so_number);

                    removeNulls(fulfillmentOrders);
                    fulfillmentOrders = sortByDateFn(fulfillmentOrders, 'definition.dueDate');

                    setServiceOrder({
                        ...servicingOrder,
                        fulfillmentOrders: showLoading(fulfillmentOrders),
                        servicingOrderItems,
                    });
                    const barcodes = getBarCodes(fulfillmentOrders);
                    fetchAssetInfo(barcodes).then(assetInfo => {
                        const newFulfillmentOrders = populateAssetInfo(fulfillmentOrders, assetInfo);
                        setServiceOrder({
                            ...servicingOrder,
                            fulfillmentOrders: newFulfillmentOrders,
                            servicingOrderItems,
                        });
                        setSelectedFulfillmentOrderID(get(newFulfillmentOrders, '[0].id', ''));
                        setSelectedOrder(newFulfillmentOrders[0]);
                    });

                    setSelectedFulfillmentOrderID(get(fulfillmentOrders, '[0].id', ''));
                } else {
                    const fulfillmentOrders = await servicingOrdersService.getFulfilmentOrdersForServiceOrder(
                        servicingOrder.so_number
                    );

                    setServiceOrder({
                        ...servicingOrder,
                        ...fulfillmentOrders,
                    });
                    /*
                    Todo : uncomment below when MGM stories are done
                    const barcodes = getBarCodes(fulfillmentOrders);
                    fetchAssetInfo(barcodes).then(assetInfo => {
                        const newFulfillmentOrders = populateAssetInfo(fulfillmentOrders, assetInfo);
                        setServiceOrder({
                            ...servicingOrder,
                            fulfillmentOrders: newFulfillmentOrders,
                        });
                        setSelectedFulfillmentOrderID(get(newFulfillmentOrders, '[0].id', ''));
                    });
                    */
                    setSelectedFulfillmentOrderID(get(fulfillmentOrders, '[0].id', ''));
                }
            } catch (e) {
                setServiceOrder(servicingOrder);
            }
        } else {
            setServiceOrder(servicingOrder);
        }
    };

    useEffect(() => {
        servicingOrdersService.getServicingOrderById(match.params.id).then(servicingOrder => {
            if (servicingOrder) {
                fetchFulfillmentOrders(servicingOrder);
            } else {
                setServiceOrder({});
            }
        });
    }, [match]);

    const handleSelectedSourceChange = source => {
        // CURRENT SELECTED SOURCE
        setSelectedSource(source);
    };

    const handleFulfillmentOrderChange = id => {
        if (selectedFulfillmentOrderID !== id) {
            setSelectedFulfillmentOrderID(id);
            setSelectedSource(null);
        }
    };

    const isFormDisabled = selectedOrder => {
        const {readiness} = selectedOrder;
        return readiness === 'READY';
    };

    return (
        <div className="servicing-order">
            <div className="servicing-order__left">
                {serviceOrder && (
                    <HeaderSection
                        orderDetails={serviceOrder}
                        handleFulfillmentOrderChange={handleFulfillmentOrderChange}
                        selectedFulfillmentOrder={selectedFulfillmentOrderID}
                    />
                )}
            </div>
            <div className="servicing-order__right">
                <FulfillmentOrder
                    selectedFulfillmentOrder={selectedOrder}
                    setSelectedOrder={setSelectedOrder}
                    setSelectedFulfillmentOrderID={setSelectedFulfillmentOrderID}
                    fetchFulfillmentOrders={fetchFulfillmentOrders}
                    serviceOrder={serviceOrder}
                    updatedServices={updatedServices}
                    cancelEditing={() => {
                        setSelectedSource({...selectedSource});
                        setSelectedOrder({...selectedOrder});
                    }}
                    lastOrder={lastOrder}
                >
                    <SourcesTable
                        onSelectedSourceChange={handleSelectedSourceChange}
                        data={prepareRowData(selectedOrder)}
                        setUpdatedServices={setUpdatedServices}
                        isDisabled={isFormDisabled(selectedOrder)}
                    />
                    {selectedSource && (
                        <ServicesTable
                            data={selectedSource}
                            isDisabled={isFormDisabled(selectedOrder)}
                            setUpdatedServices={setUpdatedServices}
                        />
                    )}
                </FulfillmentOrder>
            </div>
        </div>
    );
};

ServicingOrder.propTypes = {
    match: PropTypes.object,
};

ServicingOrder.defaultProps = {
    match: {},
};

export default ServicingOrder;
