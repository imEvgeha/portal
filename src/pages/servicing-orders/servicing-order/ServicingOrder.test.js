import React from 'react';
import {shallow} from 'enzyme';
import ServicingOrder from './ServicingOrder';
import HeaderSection from './components/header-section/HeaderSection';

describe('ServicingOrder', () => {
    describe('service section with async data fetch', () => {
        const promise = new Promise((resolve, reject) =>
            setTimeout(
                () =>
                    resolve({
                        data: {
                            configured_pr_id: 'pr_riEGi',
                            createdAt: '2020-05-25T18:09:41.614Z',
                            createdBy: null,
                            description: 'DEMO',
                            external_id: 'SR00001116',
                            id: 'so_riEGi',
                            notes: '',
                            readiness: 'NEW',
                            rush_order: '',
                            so_number: 'SO_0000000004',
                            sr_due_date: '2026-05-25T18:09:00.775Z',
                            status: 'NOT_STARTED',
                            submitted_by: '',
                            submitted_date: '',
                            tenant: 'MGM',
                            type: 'ServicingOrder',
                            updatedAt: '2020-05-25T18:09:41.614Z',
                            updatedBy: null,
                        },
                    }),
                100
            )
        );
        const wrapper = shallow(<ServicingOrder />);
        it('should render header section', () => {
            expect(wrapper.find('.servicing-order__left').length).toEqual(1);
        });
        it('should render right section', () => {
            expect(wrapper.find('.servicing-order__right').length).toEqual(1);
        });
        expect(wrapper.find(HeaderSection).length).toEqual(1);
        promise.then(() => {
            expect(wrapper.find(HeaderSection).length).toEqual(1);
        });
    });
});
