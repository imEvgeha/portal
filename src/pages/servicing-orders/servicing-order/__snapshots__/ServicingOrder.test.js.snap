// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicingOrder service section with async data fetch should match snapshot 1`] = `
<div
  className="servicing-order"
>
  <div
    className="servicing-order__left"
  >
    <Loading
      spinnerSize="large"
    />
  </div>
  <div
    className="servicing-order__right"
  >
    <Loading
      spinnerSize="large"
    />
  </div>
</div>
`;
